### DISCUSSION TITLE
我希望为Alist贡献新的onedrive驱动

### DISCUSSION THREAD

**Discussion by linshuboy** [AlistGo/alist#3902] [Q&A]
### Please make sure of the following things

- [X] I have read the [documentation](https://alist.nn.ci).
- [ ] I'm sure there are no duplicate issues or discussions.
- [X] I'm sure this feature is not implemented.
- [X] I'm sure it's a reasonable and popular requirement.

### Description of the feature / 需求描述

你好,我希望为Alist贡献新的onedrive驱动,虽然我不是go开发者,根据原有的onedrive修改一个版本的能力还是有的,但是因为这个原因,我对于一开始如何在本地跑起来代码有些疑问,希望能得到你的帮助
我做过以下尝试,通过golang初始化了本地代码,配置了goproxy,也安装了gcc,然后运行 go run main.go.提示
![image](https://user-images.githubusercontent.com/20389206/226257521-1971b206-a10f-4c82-a5ba-b46d83f2c7fe.png)
但是并没有这个进程.所以需要作者的帮助.

**Comment by welcome** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by anwen-anyi** [comment]
这不是启动吗了？ 如果你想要看带日志的那种可以使用  `server` 参数来当做启动

**Comment by linshuboy** [comment]
@anwen-anyi 是的,我又换了server方式,已经启动了,现在遇到了一个新的问题,我怎么获取admin用户呢,我运行go run main.go admin 获取的用户不能登录,感觉每次运行go run main.go 都会产生一个新的程序,并且互不干扰的感觉.

**Comment by anwen-anyi** [comment]
没问题啊
```go
PS C:\Users\安稳\Desktop\git\Alist-3.1.0\alist> go run .\main.go admin 
INFO[2023-03-20 14:10:18] reading config file: data\config.json        
INFO[2023-03-20 14:10:18] load config from env with prefix: ALIST_     
INFO[2023-03-20 14:10:18] init logrus...
INFO[2023-03-20 14:10:18] admin user's info: 
username: admin
password: w30OHV7r
```

**Reply by linshuboy** [reply]
我也使用这个命令获取了一个用户和密码，但是在server命令启动的进程中没有成功登录

**Reply by linshuboy** [reply]
所以我在想是不是因为每次运行都是重新编译连接的数据库不同造成的

**Reply by anwen-anyi** [reply]
估计没找到数据库位置，导致生成了新的. 自己排查一下吧 而且命令要和目录文件所在目录一样的

**Comment by linshuboy** [comment]
我添加的这种方式是直接授权应用访问员工数据,所以需要的配置文件不太一样,我考虑可不可以在原有的基础修改,配置的话我需要怎么向上兼容呢? 我需要添加一个认证类型,单选,默认是user 就是现在的方式,如果选择app 就是我新添加的方式,因为以前的数据是没有认证类型这个字段的,alist 对于配置文件的读取是一个什么机制

**Reply by linshuboy** [reply]
如果向上兼容不好做,我可以新开一个目录,作为这个软件的维护人员,你们建议哪一个方式

**Reply by xhofe** [reply]
建议新开一个driver。

alist的存储的配置是存储在数据库中的：
- 公共字段定义在https://github.com/alist-org/alist/blob/aeb48b2eccf5ba73d56a7ff417b2a57c6d1a49f6/internal/model/storage.go#L5
- 每个driver独有的字段定义在自己的包中，template供参考：https://github.com/alist-org/alist/blob/aeb48b2eccf5ba73d56a7ff417b2a57c6d1a49f6/drivers/template/meta.go#L8

**Comment by anwen-anyi** [comment]
这个是也是可以使用 `A1 A1P E3 E5 世纪互联`等 这些吗？

**Reply by linshuboy** [reply]
对,我是基于E5开发的,其实并不区分许可类型,只是授权方式不同,只要账户有onedrive,并授权了应用就能用onedrive_app,授权了用户,就能用onedrive,我这个理论上是支持所有环境的,只在公网,E5的账号上测试过

**Reply by linshuboy** [reply]
授权应用的方式,需要自己创建应用,不然真的没法保证数据安全

**Reply by anwen-anyi** [reply]
注册应用的时候选择 `公共客户端本机(移动和桌面)`  回调参数还是写和Web版本一样的哪个吗？  其他的不变？

**Reply by linshuboy** [reply]
使用这个需要满足这么几点:1.注册一个应用,授予应用授权文件相关权限,2.通过redirect('https://login.microsoftonline.com/common/adminConsent?client_id=' . config('microsoft.microsoft_client_id') . '&redirect_uri=' . route('onedrive.app.callback')) 应用授权网址登录管理员账号进行授权,回调域名不重要,因为不需要获取什么东西,随便找一个能访问的网址就行,把这个网址添加到应用web回调域名里面,3.然后获取应用id,应用密钥,租户id,和用户邮箱地址,填入配置应该就能用了.

**Reply by anwen-anyi** [reply]
emmm 好吧 我还是等你文档写好了再看看 完善一下吧

**Reply by linshuboy** [reply]
你是想用还是什么

**Reply by linshuboy** [reply]
可以加个V,沟通一下,用户名就是微信号

**Reply by anwen-anyi** [reply]
我就是想着测试一下顺便把文档写了 发现无从下手 哈哈

**Reply by linshuboy** [reply]
你要是能帮忙那再好不过了,文档这个东西我不在行,头大哈哈,我把流程整理好,发给你吧

**Reply by anwen-anyi** [reply]
也可以的 嘿嘿~

**Reply by linshuboy** [reply]
首先使用个人或者组织账号登录
https://entra.microsoft.com/#view/Microsoft_AAD_IAM/TenantOverview.ReactView?Microsoft_AAD_IAM_legacyAADRedirect=true
获取租户id
![image](https://user-images.githubusercontent.com/20389206/226787224-00adcbec-d412-46ff-8667-4982961519fb.png)
创建应用
![image](https://user-images.githubusercontent.com/20389206/226787858-489f1e08-26d0-4859-bb3e-5f4fd59ac035.png)
跳转链接随便填
获取应用id
![image](https://user-images.githubusercontent.com/20389206/226787990-044a9ace-207a-42af-aa45-a4ad8ae5f99d.png)
添加密钥
![image](https://user-images.githubusercontent.com/20389206/226789527-42185ceb-c52a-4912-87e6-796290f52616.png)
![image](https://user-images.githubusercontent.com/20389206/226789631-7893cd3a-f2a9-4f05-8af7-9ff45e061187.png)
![image](https://user-images.githubusercontent.com/20389206/226789719-109f4ea5-c1ed-42e9-9bb3-8c4fc58b9dfe.png)
添加权限
![image](https://user-images.githubusercontent.com/20389206/226788105-335d61ee-2925-4c42-b260-c77607576d95.png)
![image](https://user-images.githubusercontent.com/20389206/226788194-611cd2d4-6e76-4524-9474-8f9ad2fa6597.png)
![image](https://user-images.githubusercontent.com/20389206/226788234-ac140cc0-8735-4436-a05f-9ee2342e6193.png)
![image](https://user-images.githubusercontent.com/20389206/226788327-52706a9c-50d2-4418-898f-b7616ad799e4.png)
授予应用权限
![image](https://user-images.githubusercontent.com/20389206/226788432-58eb7e69-72db-4539-b4d9-463f6ede498f.png)
![image](https://user-images.githubusercontent.com/20389206/226788594-974f06cd-947d-471c-af95-bbee00f9c691.png)
如果是当前账户创建的应用,则可以在企业应用直接找到,点击授权就行
![image](https://user-images.githubusercontent.com/20389206/226789057-5bcd51d6-cd87-418f-8392-ff42f441eaba.png)
如果是不同账户创建的应用,可以通过访问https://login.microsoftonline.com/common/adminConsent?client_id=' . config('microsoft.microsoft_client_id') . '&redirect_uri=' . route('onedrive.app.callback')
两种情况都会跳转到下面
![image](https://user-images.githubusercontent.com/20389206/226788675-76ed5a7d-2031-4e0a-a374-b9028f087017.png)
![image](https://user-images.githubusercontent.com/20389206/226788720-46b6c3d2-c8ea-46e5-a29a-59ee539f2d3b.png)
应用程序授权完成
然后将租户ID,应用ID,应用秘钥填入到配置中,用户Email直接填写当前租户下有onedrive权限的账户

onedrive 初始化和新用户注册需要手动完成,自行完成账户设置

**Reply by linshuboy** [reply]
你根据这个流程试试可以吧,应该是没问题的

**Reply by anwen-anyi** [reply]
好的 晚点的时候我试试看 有了大概的流程就简单了

**Comment by anwen-anyi** [comment]
下午的时候试试看吧   刚刚在看另外一个 哈哈~刚刚搞完

**Comment by anwen-anyi** [comment]
这里是怎么填写的没看懂  `client_id` 写了客户端ID或者租户ID都不行诶  `redirect_uri` 回调参数就是注册应用时候哪个随便写的呗？
```html
https://login.microsoftonline.com/common/adminConsent?client_id=' . config('microsoft.microsoft_client_id') . '&redirect_uri=' . route('onedrive.app.callback')
```

**Reply by linshuboy** [reply]
client_id  就是应用的id 回调地址随便填一个,但是得在应用的回调配置里面 ,  
![image](https://user-images.githubusercontent.com/20389206/226817693-1ceb6555-afee-40c2-a8d7-85cb079f10e3.png)
这个common 换成租户id 这个地方忘记改了

**Reply by linshuboy** [reply]
如果你用的同一个账号创建的应用,这个连接点击按钮就自动就跳转了

**Reply by anwen-anyi** [reply]
你这个是两个链接吧？

**Reply by anwen-anyi** [reply]
直接用下面的也行好像？
```html
https://login.microsoftonline.com/common/adminConsent?client_id=客户端ID
```

**Reply by linshuboy** [reply]
是同一个链接,如果应用权限选23的话,可以用common,如果是1应该是租户id

**Reply by anwen-anyi** [reply]
我是选的3 直接无脑3了 权限多好像

**Reply by linshuboy** [reply]
更换common主要是为了限制哪一个租户下的用户,其实没影响吧

**Reply by linshuboy** [reply]
3的话 common 就行,现在可以了吗

**Reply by anwen-anyi** [reply]
你写一下完整的，我下面写的那样不对啊

```html
https://login.microsoftonline.com/租户ID/adminConsent?client_id=客户端ID' . config('microsoft.microsoft_client_id') . '&redirect_uri=注册应用时候随便写的回调URL' . route('onedrive.app.callback')
```

**Reply by anwen-anyi** [reply]
> 3的话 common 就行,现在可以了吗

用你那样的链接不行，我这样可以

```html
https://login.microsoftonline.com/common/adminConsent?client_id=客户端ID
```

**Reply by linshuboy** [reply]
[https://login.microsoftonline.com/{租户ID或者common}/adminConsent?client_id={客户端ID}&redirect_uri={注册应用时候随便写的回调URL}](https://login.microsoftonline.com/%7B%E7%A7%9F%E6%88%B7ID%E6%88%96%E8%80%85common%7D/adminConsent?client_id=%7B%E5%AE%A2%E6%88%B7%E7%AB%AFID%7D&redirect_uri=%7B%E6%B3%A8%E5%86%8C%E5%BA%94%E7%94%A8%E6%97%B6%E5%80%99%E9%9A%8F%E4%BE%BF%E5%86%99%E7%9A%84%E5%9B%9E%E8%B0%83URL%7D)

**Reply by linshuboy** [reply]
config啥的都是参数

**Reply by linshuboy** [reply]
需要替换掉的

**Reply by linshuboy** [reply]
你那样也行,不确定回调之后什么样

**Reply by anwen-anyi** [reply]
> [https://login.microsoftonline.com/{租户ID或者common}/adminConsent?client_id={客户端ID}&redirect_uri={注册应用时候随便写的回调URL}](https://login.microsoftonline.com/%7B%E7%A7%9F%E6%88%B7ID%E6%88%96%E8%80%85common%7D/adminConsent?client_id=%7B%E5%AE%A2%E6%88%B7%E7%AB%AFID%7D&redirect_uri=%7B%E6%B3%A8%E5%86%8C%E5%BA%94%E7%94%A8%E6%97%B6%E5%80%99%E9%9A%8F%E4%BE%BF%E5%86%99%E7%9A%84%E5%9B%9E%E8%B0%83URL%7D)

这样可以了  你上面写的哪个和这个完全不一样 哈哈

**Reply by anwen-anyi** [reply]
> 你那样也行,不确定回调之后什么样

问题不大随便写的点击授权后就跳到注册应用回调URL哪个页面了

**Comment by anwen-anyi** [comment]
如果是用的E5管理员操作的以上哪些行为，然后授权，还要不要每个子账号都要去登录一下下面的链接？ 还是说直接写哪个用户的邮箱就行？
```html 
https://login.microsoftonline.com/{租户ID或者common}/adminConsent?client_id={客户端ID}&redirect_uri={注册应用时候随便写的回调URL}
```

**Reply by linshuboy** [reply]
普通用户不用访问这个链接

**Reply by linshuboy** [reply]
但是如果是新的用户的话，需要登录一些onedrive初始化一下

**Reply by anwen-anyi** [reply]
例如我用E5管理员开启了25个 账号
我只需要用管理员账号登录一下最后链接授权一下
然后我就可以直接挂载这25个账号了是吗？

**Reply by anwen-anyi** [reply]
> 但是如果是新的用户的话，需要登录一些onedrive初始化一下

这个倒是知道，肯定是提前在 onedrive初始化好的 哈哈 这个不初始化不能用

**Reply by linshuboy** [reply]
是的

**Reply by anwen-anyi** [reply]
懂了 我这次拿新建E5的账号创建一下试试看 哈哈

**Reply by linshuboy** [reply]
比以前需要授权25次要方便

**Reply by anwen-anyi** [reply]
不过这样的话  嘿嘿嘿你懂得....例如给别人开的号  管理员也可以随时查看他的了吧 哈哈（只需要填写一份邮箱就可以看了这样的话）

**Reply by linshuboy** [reply]
对的，所以e5的账号还是自己用比较好

**Reply by linshuboy** [reply]
这也是我为什么说，应用要自己建的原因

**Reply by anwen-anyi** [reply]
这样方便多了，好久没登录过的用之前那种授权方式 几个月没活动的 刷新令牌就过期了，
刚看了下 刷新令牌过期的我把邮箱改了下发现能用  不用授权去

**Reply by anwen-anyi** [reply]
还有一个问题 就是怎么看是不是管理员？
我刚刚在 https://admin.microsoft.com/Adminportal/Home?source=applauncher#/users 这里创建E5登录这里看了下没看出哪个是管理员诶~

如果用 https://admin.microsoft.com/Adminportal/Home?source=applauncher#/users 这个账号而不是用 E5域的哪个账号去新建似乎不能在最后 权限哪里授权同样

**Reply by linshuboy** [reply]
这个我都是用第一个默认创建的用户作为管理员的

**Reply by linshuboy** [reply]
其他的我也没分配过,靠人脑记?哈哈

**Comment by anwen-anyi** [comment]
在 https://entra.microsoft.com/#view/Microsoft_AAD_IAM/TenantOverview.ReactView?Microsoft_AAD_IAM_legacyAADRedirect=true 登录新建E5账号的账号打开所有应用程序提示错误知道怎么回事吗？
![image](https://user-images.githubusercontent.com/56105412/226826828-8b64971c-4e42-4168-818d-2b1a2db25427.png)

**Reply by linshuboy** [reply]
不知道,,我从来都不登录成员账号

**Reply by anwen-anyi** [reply]
这个是E5域账号的祖宗 哈哈，没这个账号就没E5域账号...

**Reply by anwen-anyi** [reply]
没事儿了QAQ...知道怎么看了

**Comment by anwen-anyi** [comment]
回调参数 不写 https://entra.microsoft.com/TokenAuthorize 倒数第二步的时候授权不了好像 哈哈

**Reply by linshuboy** [reply]
这个应该没问题,我的回调参数写的是我自己的网站

**Reply by anwen-anyi** [reply]
emmm 我也不知道咋回事，反正下面序号的第四步授权哪儿 一直提示回调参数有问题 我改成了它给的哪个 就好了反正 哈哈

**Reply by linshuboy** [reply]
还有新的E5账号吗,要不要帮我测试一个东西

**Reply by anwen-anyi** [reply]
就一个E5账号QAQ，测试什么

**Reply by linshuboy** [reply]
一键初始化,E5账号,我想看看稳定性,我再注册吧

**Reply by linshuboy** [reply]
你的E5是怎么保活的

**Reply by linshuboy** [reply]
我想起来了,如果是通过界面授权的,确实需要添加这个回调

**Reply by anwen-anyi** [reply]
MS365 E5 Renew 和 https://e5.qyi.io/ 还有AList

**Reply by anwen-anyi** [reply]
> 我想起来了,如果是通过界面授权的,确实需要添加这个回调

哈哈 我预判了你的预判~

**Comment by anwen-anyi** [comment]
写好了 你可以预览一下先 看看哪里不合适一起改了QAQ
https://deploy-preview-187--alist-doc.netlify.app/guide/drivers/onedrive_app.html

**Reply by linshuboy** [reply]
完美,不用改了,比我整理的要详细

**Comment by anwen-anyi** [comment]
世纪互联 这是啥错误哩~  我也远程看了看是全局管理员 也授权了~
`Failed init storage: AADSTS500011: The resource principal named https://graph.microsoft.com/ was not found in the tenant named xxxxxxxx. This can happen if the application has not been installed by the administrator of the tenant or consented to by any user in the tenant. You might have sent your authentication request to the wrong tenant. Trace ID: bb21xxxxxxxxxx5a7a00 Correlation ID: ba7c0xxxxxxxxxxxx3539800a23 Timestamp: 2023-03-28 12:34:17Z`

**Reply by linshuboy** [reply]
这个解决了吗，好像是管理员并没有安装这个应用

**Reply by anwen-anyi** [reply]
> 这个解决了吗，好像是管理员并没有安装这个应用

不清楚这个是 `世纪互联` 操作的具体怎么搞我也不是很清楚了..暂时没管了

**Reply by linshuboy** [reply]
世纪互联和国际版是不是不互通，我没具体测试过，抽空看看吧

**Reply by anwen-anyi** [reply]
> 世纪互联和国际版是不是不互通，我没具体测试过，抽空看看吧

没管了 `世纪互联` 好像也在开始抛弃国内业务了

**Comment by FoundZiGu** [comment]
存储驱动OneDrive APP，授权啥的步骤全部按要求操作了，也都成功了，最后不能显示文件夹

**Reply by linshuboy** [reply]
不能显示文件夹报什么错误

**Reply by timorpic** [reply]
> 不能显示文件夹报什么错误

failed get objs: failed to list objs: User not found
